apps/meteor/server/email/IMAPInterceptor.ts

```diff
@@ -49,132 +51,186 @@ export class IMAPInterceptor extends EventEmitter {
+					resolve(results);
+				}
+			};
+			this.imap.search(this.options.filter, cb);
+		});
+	}
 
-				fetch.on('message', (msg, seqno) => {
-					logger.debug('E-mail received', seqno, msg);
+	parseEmails(stream: NodeJS.ReadableStream, _info: ImapMessageBodyInfo): Promise<ParsedMail> {
+		return new Promise((resolve, reject) => {
+			const cb = (err: Error, mail: ParsedMail) => {
+				if (err) {
+					reject(err);
+				} else {
+					resolve(mail);
+				}
+			};
+			simpleParser(stream, cb);
+		});
+	}
 
-					msg.on('body', (stream, type) => {
-						if (type.which !== '') {
+	imapFetch(emailIds: number[]): Promise<number[]> {
+		return new Promise((resolve, reject) => {
+			const out: number[] = [];
+			const messagecb = (msg: ImapMessage, seqno: number) => {
+				out.push(seqno);
+				const bodycb = (stream: NodeJS.ReadableStream, _info: ImapMessageBodyInfo): void => {
+					simpleParser(stream, (_err, email) => {
+						if (this.options.rejectBeforeTS && email.date && email.date < this.options.rejectBeforeTS) {
+							logger.error(`Rejecting email on inbox ${this.config.user}`, email.subject);
 							return;
 						}
-
-						simpleParser(stream, (_err, email) => {
-							if (this.options.rejectBeforeTS && email.date && email.date < this.options.rejectBeforeTS) {
-								logger.error(`Rejecting email on inbox ${this.config.user}`, email.subject);
-								return;
-							}
-							this.emit('email', email);
-						});
+						this.emit('email', email);
 					});
+				};
+				msg.once('body', bodycb);
+			};
+			const errorcb = (err: Error): void => {
+				logger.warn(`Fetch error: ${err}`);
+				reject(err);
+			};
+			const endcb = (): void => {
+				resolve(out);
+			};
+			const fetch = this.imap.fetch(emailIds, {
+				bodies: ['HEADER', 'TEXT', ''],
+				struct: true,
+				markSeen: this.options.markSeen,
+			});
+
+			fetch.on('message', messagecb);
+			fetch.on('error', errorcb);
+			fetch.on('end', endcb);
+		});
+	}
 
-					// On fetched each message, pass it further
-					msg.once('end', () => {
-						// delete message from inbox
-						if (this.options.deleteAfterRead) {
-							this.imap.seq.addFlags(seqno, 'Deleted', (err) => {
-								if (err) {
-									logger.error(`Mark deleted error: ${err}`);
-								}
-							});
-						}
-					});
-				});
-
-				fetch.once('error', (err) => {
-					logger.error(`Fetch error: ${err}`);
+	// Fetch all UNSEEN messages and pass them for further processing
+	async getEmails(): Promise<void> {
+		const emailIds = await this.imapSearch();
+		const emailsFetched = await this.imapFetch(emailIds);
+
+		// this.imapSearch()
+		// 	.then(this.imapFetch)
+		// 	.then((emails) => {
+		// emails.forEach((email) => {
+		for (const email of emailsFetched) {
+			// this.emit('email', email.mail);
+			if (this.options.deleteAfterRead) {
+				this.imap.seq.addFlags(email, 'Deleted', (err) => {
+					if (err) {
+						logger.warn(`Mark deleted error: ${err}`);
+					}
 				});
 			}
-		});
+		}
+	}
```

Comment:
Noticed this is refetching emails 🤔 while the prev implementation was attaching a listener to "end" of each message, meaning it's using the same already fetched data. Any reason to change that behavior?